devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,256 @@
+
+        self._deriv_order = deriv_order
+        self._function = function
+        self._dimension = dimension
+        self._coefficients = coefficients
+
+    @property
+    def deriv_order(self):
+        """The derivative order."""
+        return self._deriv_order
+
+    @property
+    def function(self):
+        """The function to which the coefficients belong."""
+        return self._function
+
+    @property
+    def dimension(self):
+        """The dimension to which the coefficients will be applied."""
+        return self._dimension
+
+    @property
+    def coefficients(self):
+        """The set of coefficients."""
+        return self._coefficients
+
+    def check_input(self, deriv_order, function, dimension, coefficients):
+        if not isinstance(deriv_order, int):
+            raise TypeError("Derivative order must be an integer")
+        # NOTE: Can potentially be tidied up following the implementation
+        # of lazy evaluation.
+        try:
+            if not function.is_Function:
+                raise TypeError("Object is not of type Function")
+        except:
+            raise TypeError("Object is not of type Function")
+        try:
+            if not dimension.is_Dimension:
+                raise TypeError("Coefficients must be attached to a valid dimension")
+        except:
+            raise TypeError("Coefficients must be attached to a valid dimension")
+        # Currently only numpy arrays are accepted here.
+        # Functionality will be expanded in the near future.
+        if not isinstance(coefficients, np.ndarray):
+            raise NotImplementedError
+        return
+
+
+class CoefficientRules(object):
+    """
+    Devito class to convert Coefficient objects into replacent rules
+    to be applied when constructing a Devito Eq.
+
+    Example
+    -------
+    Assume we have created the 'u_x_coeffs' Coefficient object
+    as shown above in the Class Coefficient example.
+
+    >>> from devito import CoefficientRules
+    >>> coeffs = CoefficientsRules(u_x_coeffs)
+
+    Now create a Devito equation and pass to it 'coeffs'
+
+    >>> from devito import Eq
+    >>> eq = Eq(u.dt+u.dx, coefficients=coeffs)
+    Eq(0.1*u(t, x, y) - 0.6*u(t, x - h_x, y) + 0.6*u(t, x + h_x, y)
+    - u(t - dt, x, y)/(2*dt) + u(t + dt, x, y)/(2*dt), 0)
+
+    Notes
+    -----
+    If a function is declared with 'symbolic' coefficients and no
+    replacement rules for any derivative appearing in a Devito equation,
+    the coefficients will revert to those of the 'default' Taylor
+    expansion.
+    """
+
+    def __init__(self, *args):
+
+        if any(not isinstance(arg, Coefficient) for arg in args):
+            raise TypeError("Non Coefficient object within input")
+
+        self._data = args
+        self._function_list = self.function_list
+        self._rules = self.rules
+
+    @property
+    def data(self):
+        """The Coefficient objects passed."""
+        return self._data
```

Comment:
for code self-documentation: why not "coefficients", instead of ``data``?

Reply:
It was for naming reasons to avoid things such as `coeffs.coefficients`. I've changed it to `coefficients` for now and will have a re-look later to see if the various names need some re-organising.